resolwe_bio/processes/import_data/metadata.py

```diff
@@ -1,93 +1,161 @@
 
-def change_suffix(path):
-    """Change suffix of a file to lowercase."""
-    new_path = path.with_suffix(path.suffix.lower())
-    path.replace(new_path)
-    return new_path
+    def run(self, inputs, outputs):
+        """Run the analysis."""
 
+        collections = Collection.filter(data__id=self.data.id)
+        if not collections:
+            self.error(
+                "Metadata table was not uploaded to a Collection. "
+                "Matching of metadata entries to Sample objects is not possible."
+            )
 
-class UploadOrangeMetadata(Process):
-    """Upload metadata table in Orange format.
+        path = Path(inputs.src.import_file(imported_format="extracted"))
 
-    Orange can read files in native tab-delimited format, or can load
-    data from any of the major standard spreadsheet file types, like CSV
-    and Excel. Native format starts with the names of attributes with
-    prefixes that define attribute type (continuous, discrete, time,
-    string) and role (class, meta, ignore, instance weights). Prefixes
-    are separated from the attribute name with a hash sign (“#”).
+        # change the file suffix if it is either upper or mixed case
+        if not path.suffix.islower():
+            path = change_suffix(path)
+            self.info(
+                "File extension of the table was replaced with a lower case version."
+            )
+
+        try:
+            df_data = read_tabular_data(path, error=self.error)
+        except Exception as err:
+            self.error(f"It was not possible to read the provided data table. {err}")
+
+        sample_columns = {
+            "Sample ID": "id",
+            "Sample slug": "slug",
+            "Sample name": "name",
+        }
+
+        sample_headers = df_data.columns.intersection(sample_columns)
+        if len(sample_headers) != 1:
+            self.error(
+                f"The uploaded metadata table needs to contain "
+                f"exactly one of the following columns: "
+                f"{sorted(sample_columns.keys())}."
+            )
+
+        n_samples = len(df_data[sample_headers[0]].unique())
+        if n_samples < 1:
+            self.error("The uploaded table contains no samples.")
+
+        samples = Entity.filter(collection_id=collections[0].id)
 
-    Legacy format with three header rows is also supported. The first
-    row lists feature (column) names. The second header row gives the
-    attribute type and the third header line contains role information
-    to identify dependent features (class), irrelevant features (ignore)
-    or meta features (meta).
+        col_samples = {
+            getattr(sample, sample_columns[sample_headers[0]]) for sample in samples
+        }
 
-    For more information see Orange
-    [documentation](https://orange-visual-programming.readthedocs.io/loading-your-data/index.html#header-with-attribute-type-information).
+        df_samples = df_data[sample_headers[0]].values
```

Comment:
If you define `df_samples = df_data[sample_headers[0]]` and use them one line lower as 
`intersection = col_samples.intersection(df_samples.values)` you can tham omit creating the `sample_s` variable. This would simplify code and also remove one of the sample-something variable that makes code a bit hairder to digest.